### PR TITLE
Additional validation on malformed images

### PR DESCRIPTION
**Pull request by lucab**
This PR introduces some additional validation before converting Docker images. This addresses:
 * an infinite loop when converting images with a cyclic dependency chain (CVE-2016-8579)
 * a path traversal when extracting layers with crafted IDs (CVE-2016-7569)

It also introduces fixtures and tests script for such cases.

Fixes https://github.com/appc/docker2aci/issues/201
Fixes https://github.com/appc/docker2aci/issues/203

### CODE REVIEW

File: lib/internal/backend/file/file.go [hunk 1]
```diff
@@ -279,14 +282,24 @@ func extractEmbeddedLayer(file *os.File, layerID string, outputPath string) (*os
...
+// chain and errors out in such cases.
 func getAncestry(file *os.File, imgID string) ([]string, error) {
 	var ancestry []string
+	deps := make(map[string]bool)
 
 	curImgID := imgID
 
 	var err error
 	for curImgID != "" {
+		if _, ok := deps[curImgID]; ok {
```
Comment (jonboulle) [inline lib/internal/backend/file/file.go]: nit, since you're using a bool you can just use the default value check (`if deps[curImgID]`)

**Comment by lucab** [comment]
@jonboulle rebased to address the nit and to include the reference to the second CVE.

/cc @TheBeeMan

File: lib/common/common.go [hunk 2]
```diff
@@ -60,3 +63,21 @@ func ParseDockerURL(arg string) (*ParsedDockerURL, error) {
 	p, err := docker.ParseDockerURL(arg)
 	return (*ParsedDockerURL)(p), err
 }
+
+// IsValidLayerRefs validates a layer refs
```
Comment (jonboulle) [inline lib/common/common.go]: unused - remove this?

File: lib/internal/backend/file/file.go [hunk 3]
```diff
@@ -279,14 +282,24 @@ func extractEmbeddedLayer(file *os.File, layerID string, outputPath string) (*os
 	return layerFile, nil
 }
 
+// getAncestry computes an image ancestry, returning an ordered list
+// of dependencies starting from the topmost image to the base.
+// It checks for dependency loops via duplicates detection in the image
```
Comment (jonboulle) [inline lib/internal/backend/file/file.go]: duplicate detection

File: lib/internal/backend/file/file.go [hunk 4]
```diff
@@ -279,14 +282,24 @@ func extractEmbeddedLayer(file *os.File, layerID string, outputPath string) (*os
...
 	curImgID := imgID
 
 	var err error
 	for curImgID != "" {
+		if deps[curImgID] {
+			return nil, fmt.Errorf("dependency loop detected at image %q", curImgID)
+		}
+		deps[curImgID] = true
 		ancestry = append(ancestry, curImgID)
+		log.Debug(`Getting ancestry for layer "`, curImgID, `".`)
```
Comment (jonboulle) [inline lib/internal/backend/file/file.go]: IMO nicer to just format the string here: `fmt.Sprintf("... %q ...", curImgID)`

File: lib/common/common.go [hunk 5]
```diff
@@ -60,3 +63,21 @@ func ParseDockerURL(arg string) (*ParsedDockerURL, error) {
...
+func IsValidLayerRefs(name string) error {
+	validName := regexp.MustCompile(`^([A-Fa-f0-9_+.-]+)$`)
+	if ok := validName.MatchString(name); !ok {
+		return fmt.Errorf("invalid layer name %q", name)
+	}
+	return nil
+}
+
+// IsValidLayerId validates a layer ID
```
Comment (jonboulle) [inline lib/common/common.go]: Naming nit: generally `IsFoo` functions return a `bool`. Instead perhaps `AssertValidLayerId` or `ValidateLayerID`

Comment (lucab) [inline lib/common/common.go]: My bad, this was just returning a bool at first but then changed to error returning, but I forgot to reflect that in the naming.

File: lib/common/common.go (hunk 2 above)
Comment (lucab) [inline lib/common/common.go]: This was supposed to be a library-exposed helper, but I guess we may re-introduce it later as soon as any consumers need it.

**Comment by lucab** [comment]
Further iteration on this: removed unused bits, renamed function and cleaned the debug printing.

File: lib/common/common.go [hunk 6]
```diff
@@ -60,3 +63,12 @@ func ParseDockerURL(arg string) (*ParsedDockerURL, error) {
 	p, err := docker.ParseDockerURL(arg)
 	return (*ParsedDockerURL)(p), err
 }
+
+// ValidateLayerId validates a layer ID
+func ValidateLayerId(id string) error {
+	validId := regexp.MustCompile(`^(\w+:)?([A-Fa-f0-9]+)$`)
```
Comment (jonboulle) [inline lib/common/common.go]: Sorry, missed this on previous pass: this should a global variable (so failure would be at initialization time, not function invocation time)

Comment (lucab) [inline lib/common/common.go]: Ack, moved to a global var stanza.

**Pull request merged**